# Простые типы данных

In [ ]:
my_integer = 10
type(my_integer)

In [ ]:
my_float = 5.5
type(my_float)

In [ ]:
my_string = 'Hello World!'
my_string_2 = "Hello World"
type(my_string)

In [ ]:
# my_bool = True
my_bool = False
type(my_bool)

In [ ]:
x = 5
y = 1
type(x > y)

In [ ]:
# преобразование типов
salary = 1000
print('Ваша годовая зарплата составляет ' + salary + ' условных единиц')

In [ ]:
print('Ваша годовая зарплата составляет ' + str(salary) + ' условных единиц')

In [ ]:
# неявное преобразование типов
# print(1 + False)
# print(20 / 5.0)

## Строки

In [ ]:
'Hi, ' + 'Oleg'

In [ ]:
my_string.upper()

In [ ]:
my_string.lower()

In [ ]:
my_string.capitalize()

In [ ]:
my_string.replace('Hello', 'Goodbye')

In [ ]:
len(my_string)

## Индексация и срезы

In [2]:
my_string = 'Hello World'

In [3]:
my_string[2]

'l'

In [ ]:
my_string[-4]

In [ ]:
my_string[0:5]

In [ ]:
my_string = 'Hello World'
my_string[0:8:2]

In [ ]:
my_string[6:]

In [1]:
my_string[:5]

NameError: name 'my_string' is not defined

## Упражнение
Выделите номер часа из даты следующего формата

In [ ]:
date = '2019-08-27T23:59:00.932'

## Проверка на вхождение элемента в объект

In [ ]:
my_string = 'Hello World'
target_string = 'World'
if target_string not in my_string:
    print('not find!')

## f-строки

In [4]:
name = 'oleg'
lang = 'python'
t = f"Hello, {name.capitalize()}, i know {lang} a bit"
print(t)
print(type(t))

Hello, Oleg, i know python a bit
<class 'str'>


# Списки

In [ ]:
month_list = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep']
income_list = [13000, 14000, 14300, 15000, 13800, 13000, 14900, 15200, 15300]
income_by_months = [['Jan', 13000], ['Feb', 14000], ['Mar', 14300], ['Apr', 15000], ['May', 13800], ['Jun', 13000], ['Jul', 14900], ['Aug', 15200], ['Sep', 15300]]

In [ ]:
print(type(month_list))
print(type(income_list))
print(type(income_by_months))

In [ ]:
# индексация элементов в списке
print(month_list[0])
print(month_list[-1])
print(income_by_months[-4])

In [ ]:
# срезы
print(income_by_months[0:2])
print('--------------')
print(income_by_months[-8:-6])
print('--------------')
print(income_by_months[2:])
print('--------------')
print(income_by_months[:3])

In [ ]:
# можно обращаться к любому уровню вложенности
income_by_months = [['Jan', 13000], ['Feb', 14000], ['Mar', 14300], ['Apr', 15000], ['May', 13800], ['Jun', 13000], ['Jul', 14900], ['Aug', 15200], ['Sep', 15300]]
income_by_months[0][0]

In [ ]:
# изменение списков
income_by_months[0][1] = 13100
print(income_by_months)

In [ ]:
income_by_months[0:2] = [['Jan', 13200], ['Feb', 13900]]
print(income_by_months)

In [ ]:
income_by_months_2 = [['Nov', 15400], ['Dec', 17000]]
income_by_month = income_by_months + income_by_months_2
print(income_by_month)

## Распаковка списков

In [ ]:
first, second, third = ['первый', 'второй', 'третий']

In [ ]:
# когда число элементов неизвестно
first, *other = ['первый', 'второй', 'третий']
other

In [ ]:
first, *other, last =  ['первый', 'второй', 'третий', 'четвертый']

In [ ]:
first, last

## Операции со списками

In [ ]:
income_by_months

In [ ]:
# Удаляем элемент по индексу
del(income_by_months[-1])
income_by_months

In [ ]:
# удаляем элемент по значению
month_list.remove('Sep')
print(month_list)

In [ ]:
# добавляем элемент в конец списка
income_by_months.append(['Dec', 17000])
income_by_months

In [ ]:
# добавляем элемент по нужному индексу
income_list.insert(2, 1111111)
print(income_list)

In [ ]:
# считаем количество вхождений элемента в список
income_list.count(13000)

In [ ]:
# узнаем индекс элемента в списка (только первое вхождение!)
income_list.index(13000, 1)

In [ ]:
# разворачиваем список
month_list.reverse()
month_list

In [ ]:
# узнаем длину списка
len(income_list)

In [ ]:
# сумма элементов
sum(income_list)

In [ ]:
# максимальный элемент элементов
max(income_list)

In [ ]:
# минимальный элемент элементов
min(income_list)

In [ ]:
# сортировка по возрастанию
sorted(income_list)

In [ ]:
# изменить порядок сортировки
sorted(income_list, reverse=True)

In [ ]:
# а это сортировка строк по алфавиту
sorted(month_list)

## Изменение списков

В примере ниже переменная a и b на самом деле указывают на один и тот же объект. В результате, при добавлении в b очередного элемента этот элемент добавляется и в исходном листе a

In [ ]:
a = [1, 2, 3]
b = a

In [ ]:
b.append(4)
f'a = {a}'

In [ ]:
id(a), id(b)

Используем модуль copy

In [5]:
import copy

In [6]:
a = [1, 2, 3]
b = copy.copy(a)

In [7]:
id(a), id(b)

(2621024705280, 2621024768576)

In [8]:
b.append(4)
print('a = {}'.format(a))

print('b = {}'.format(b))

a = [1, 2, 3]
b = [1, 2, 3, 4]


## Списки и строки

In [ ]:
queries_string = "смотреть сериалы онлайн,новости спорта,афиша кино,курс доллара,сериалы этим летом,курс по питону,сериалы про спорт"

In [ ]:
# преобразование строки в список (например, из CSV-файла)
queries_string.split(',')

In [ ]:
# Преобразование списка в строку

','.join(['Столбец 1', 'Столбец 2', 'Столбец 3'])

In [ ]:
# проверка вхождения элемента в список:

# 'Москва' in ['Ленинград', 'Одесса', 'Севастополь', 'Москва']
'Москва' not in ['Ленинград', 'Одесса', 'Севастополь', 'Москва']

# Tuple (кортежы)

In [ ]:
salary_tuple = (1000, 1200, 1300, 900, 800)
type(salary_tuple)
# type(list(salary_tuple))

In [ ]:
# print(salary_tuple[0])
salary_tuple[0] = 500

In [ ]:
# кортеж из одного элемента задается так:
t = ('one', )
type(t)

In [ ]:
# без запятой получится строка
type( ('one') )

In [ ]:
# функция zip
salaries = [1000, 1200, 1300, 900, 800, 1000]
names = ['Robert', 'Jane', 'Liza', 'Richard', 'John']
salaries_by_names = zip(names, salaries)
# print(salaries_by_names)
print(list(salaries_by_names))

# Циклы

## Цикл while

In [ ]:
x = 5
while x != 0:
    x -= 1
    print(x)
    

In [ ]:
x = 6
while x != 0:
    if x % 2 == 0:
        print(x, '- четное число')
    else:
        print(x, '- нечетное число')
    x -= 1

## Цикл for

In [ ]:
# итерация по строкам
company_name = 'Orange'
for letter in company_name:
#     print(letter)
    letter = letter.capitalize()
    print(letter)

In [ ]:
# итерация по спискам
companies_capitalization = [
 ['Orange', 1.3],
 ['Maxisoft', 1.5],
 ['Headbook', 0.8],
 ['Nicola', 2.2]
]

for company in companies_capitalization:
#     print(company)
    print(company[0], 'capitalization is', company[1])
#     print('end of iteration')

### Функции range и enumerate

In [ ]:
# range(10)
type(range(10))

In [ ]:
list(range(1, 10))

In [ ]:
for i in range(10):
    print(i)

In [ ]:
# с указанием левой и правой границы
for i in range(3, 20):
    print(i)

In [ ]:
# третий аргумент - шаг
for i in range(3, 20, 5):
    print(i)

In [ ]:
# enumerate позволяет получить индекс каждого элемента
enumerate([144, 244, 344, 444, 544])
list(enumerate([144, 244, 344, 444, 544]))

In [ ]:
companies_capitalization = [
 ['Orange', 1.3],
 ['Maxisoft', 1.5],
 ['Headbook', 0.8],
 ['Nicola', 2.2]
]
for i, company in enumerate(companies_capitalization):
    print(i+1, company[0], 'capitalization is', company[1])

## break, pass, continue

In [ ]:
phrase = '640Кб должно хватить для любых задач. Билл Гейтс (по легенде)'

In [ ]:
for letter in phrase:
    if letter == ' ':
        break
    print(letter, end='')

In [ ]:
for letter in phrase:
    if letter == ' ':
        continue
    print(letter, end='')

In [ ]:
for letter in phrase:
    if letter == ' ':
        pass
    print(letter, end='')

## Попрактикуемся

Имеется структура данных cook_book, в которой хранится информация об ингредиентах блюд и их количестве в расчете на одну порцию и переменная, в которой хранится количество людей, на которых необходимо приготовить данные блюда:

In [10]:
cook_book = [
  ['салат',
      [
        ['картофель', 100, 'гр.'],
        ['морковь', 50, 'гр.'],
        ['огурцы', 50, 'гр.'],
        ['горошек', 30, 'гр.'],
        ['майонез', 70, 'мл.'],
      ]
  ],
  ['пицца',  
      [
        ['сыр', 50, 'гр.'],
        ['томаты', 50, 'гр.'],
        ['тесто', 100, 'гр.'],
        ['бекон', 30, 'гр.'],
        ['колбаса', 30, 'гр.'],
        ['грибы', 20, 'гр.'],
      ],
  ],
  ['фруктовый десерт',
      [
        ['хурма', 60, 'гр.'],
        ['киви', 60, 'гр.'],
        ['творог', 60, 'гр.'],
        ['сахар', 10, 'гр.'],
        ['мед', 50, 'мл.'],  
      ]
  ]
]


In [36]:
person = 5

print(f"{cook_book[0][0].capitalize()}:\n"
      f"{cook_book[0][1][0][0].capitalize()}, {cook_book[0][1][0][1] * person }{cook_book[0][1][0][2]}\n"
      f"{cook_book[0][1][1][0].capitalize()}, {cook_book[0][1][1][1] * person }{cook_book[0][1][1][2]}\n"
      f"{cook_book[0][1][2][0].capitalize()}, {cook_book[0][1][2][1] * person }{cook_book[0][1][2][2]}\n"
      f"{cook_book[0][1][3][0].capitalize()}, {cook_book[0][1][3][1] * person }{cook_book[0][1][3][2]}\n"
      f"{cook_book[0][1][4][0].capitalize()}, {cook_book[0][1][4][1] * person }{cook_book[0][1][4][2]}\n"
      f"\n"
      f"{cook_book[1][0].capitalize()}:\n"
      f"{cook_book[1][1][0][0].capitalize()}, {cook_book[1][1][0][1] * person }{cook_book[1][1][0][2]}\n"
      f"{cook_book[1][1][1][0].capitalize()}, {cook_book[1][1][1][1] * person }{cook_book[1][1][1][2]}\n"
      f"{cook_book[1][1][2][0].capitalize()}, {cook_book[1][1][2][1] * person }{cook_book[1][1][2][2]}\n"
      f"{cook_book[1][1][3][0].capitalize()}, {cook_book[1][1][3][1] * person }{cook_book[1][1][3][2]}\n"
      f"{cook_book[1][1][4][0].capitalize()}, {cook_book[1][1][4][1] * person }{cook_book[1][1][4][2]}\n"
      f"{cook_book[1][1][5][0].capitalize()}, {cook_book[1][1][5][1] * person }{cook_book[1][1][5][2]}\n"
      f"\n"
      f"{cook_book[2][0].capitalize()}:\n"
      f"{cook_book[2][1][0][0].capitalize()}, {cook_book[2][1][0][1] * person }{cook_book[2][1][0][2]}\n"
      f"{cook_book[2][1][1][0].capitalize()}, {cook_book[2][1][1][1] * person }{cook_book[2][1][1][2]}\n"
      f"{cook_book[2][1][2][0].capitalize()}, {cook_book[2][1][2][1] * person }{cook_book[2][1][2][2]}\n"
      f"{cook_book[2][1][3][0].capitalize()}, {cook_book[2][1][3][1] * person }{cook_book[2][1][3][2]}\n"
      f"{cook_book[2][1][4][0].capitalize()}, {cook_book[2][1][4][1] * person }{cook_book[2][1][4][2]}\n")



Салат:
Картофель, 500гр.
Морковь, 250гр.
Огурцы, 250гр.
Горошек, 150гр.
Майонез, 350мл.

Пицца:
Сыр, 250гр.
Томаты, 250гр.
Тесто, 500гр.
Бекон, 150гр.
Колбаса, 150гр.
Грибы, 100гр.

Фруктовый десерт:
Хурма, 300гр.
Киви, 300гр.
Творог, 300гр.
Сахар, 50гр.
Мед, 250мл.



Необходимо вывести пользователю список покупок необходимого количества ингредиентов для приготовления блюд на определенное число персон в следующем виде:

Салат:  
картофель, 500гр.  
морковь, 250гр.  
огурцы, 250гр.  
горошек, 150гр.  
майонез, 350мл.  

Пицца:  
сыр, 250гр.  
томаты, 250гр.  
тесто, 500гр.  
бекон, 150гр.  
колбаса, 150гр.  
грибы, 100гр.  

Фруктовый десерт:  
хурма, 300гр.  
киви, 300гр.  
творог, 300гр.  
сахар, 50гр.  
мед, 250мл.  

## List comprehension

In [37]:
# Дана последовательность чисел. Мы хотим оставить только те, что делятся на 5
sequence = range(0, 40, 3)
list(sequence)

[0, 3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39]

In [38]:
# решение в лоб
for num in sequence:
    if num % 5 == 0:
        print(num)

0
15
30


In [39]:
# если хотим получить отфильтрованный лист, то будет даже так
filtered_sequence = []

for num in sequence:
    if num % 5 == 0:
        filtered_sequence.append(num)

print(filtered_sequence)

[0, 15, 30]


In [40]:
[num for num in sequence if num % 5 == 0]

[0, 15, 30]

In [41]:
[w for w in sequence if w % 5 == 0]

[0, 15, 30]

In [42]:
# с list comprehension получается покороче

[x**2 for x in sequence if x % 5 == 0]

[0, 225, 900]

Пример вычисления метрики из набора списков. Столбцы в каждой строке:
- дата
- номер счетчика
- количество визитов

Найдем среднее количество визитов по нашим данным

In [43]:
api_response = [
    ['2017-12-26', '777', 184],
    ['2017-12-27', '111', 146],
    ['2017-12-28', '777', 98],
    ['2017-12-29', '777', 206],
    ['2017-12-30', '111', 254],
    ['2017-12-31', '777', 89],
    ['2018-01-01', '111', 54],
    ['2018-01-02', '777', 68],
    ['2018-01-03', '777', 74],
    ['2018-01-04', '111', 89],
    ['2018-01-05', '777', 104],
    ['2018-01-06', '777', 99],
    ['2018-01-07', '777', 145],
    ['2018-01-08', '111', 184],
]

In [44]:
for element in api_response:
    print(element[2])

184
146
98
206
254
89
54
68
74
89
104
99
145
184


In [45]:
sum([x[2] for x in api_response])/len(api_response)

128.14285714285714

## Попрактикуемся

Имеется поток данных о ценах товаров (считайте, что цена всегда больше 0). Необходимо написать алгоритм определения минимального четного числа в этом потоке.

In [47]:
# sys_stdin сейчас просто название переменной
sys_stdin = [78, 68, 484, 3, 254, 90, 143, 78, 43, 42, 3053, 473, 5, 8593, 16, 3, 1454, 37, 96, 8547]

In [48]:
min(_ for _ in sys_stdin)

3

## Множества (set)
Набор неповторяющихся элементов в случайном порядке

In [50]:
data_scientist_skills = set(['Python', 'R', 'SQL', 'Tableau', 'SAS', 'Git'])
data_engineer_skills = set(['Python', 'Java', 'Scala', 'Git', 'SQL', 'Hadoop'])

In [51]:
# логическое ИЛИ – что нужно знать data-scientst, который по совместительству data-engineer
print(data_scientist_skills.union(data_engineer_skills))
print(data_scientist_skills | data_engineer_skills)

{'SAS', 'SQL', 'Tableau', 'Git', 'Scala', 'Java', 'R', 'Hadoop', 'Python'}
{'SAS', 'SQL', 'Tableau', 'Git', 'Scala', 'Java', 'R', 'Hadoop', 'Python'}


In [52]:
# логическое И – что нужно знать и data-scientist и data-engineer
print(data_scientist_skills.intersection(data_engineer_skills))
print(data_scientist_skills & data_engineer_skills)

{'SQL', 'Python', 'Git'}
{'SQL', 'Python', 'Git'}


In [53]:
# разность множеств – что знает data-scientist, но не знает data-engineer (и наоборот)
# print(data_scientist_skills.difference(data_engineer_skills))
# print(data_scientist_skills - data_engineer_skills)
print(data_engineer_skills.difference(data_scientist_skills))
print(data_engineer_skills - data_scientist_skills)

{'Scala', 'Java', 'Hadoop'}
{'Scala', 'Java', 'Hadoop'}


In [54]:
# симметричная разность множеств – что такого знают data-scientist и data-engineer, чего не знают они оба
# print(data_scientist_skills.symmetric_difference(data_engineer_skills))
# print(data_scientist_skills ^ data_engineer_skills)
print(data_engineer_skills.symmetric_difference(data_scientist_skills))
print(data_engineer_skills ^ data_scientist_skills)

{'SAS', 'Tableau', 'Scala', 'Java', 'R', 'Hadoop'}
{'SAS', 'Tableau', 'Scala', 'Java', 'R', 'Hadoop'}


In [ ]:
# Из списка можно убрать все повторения просто обратив его в set!

# Словари

In [55]:
salaries = {
    'John': 1200,
    'Mary': 500,
    'Steven': 1000,
    'Liza': 1500
}

In [56]:
# обращение к элементу словаря
salaries['John']

1200

In [57]:
# удаляем элемент из словаря
del(staff_dict['Liza'])
staff_dict

NameError: name 'staff_dict' is not defined

In [58]:
# добавляем элемент в словарь
salaries['Mary'] = 2000
salaries

{'John': 1200, 'Mary': 2000, 'Steven': 1000, 'Liza': 1500}

In [59]:
# безопасно получаем значение по ключу
salaries['Oleg']
# salaries.get('Oleg', 'Not Found')
salaries.get('Mary', 'Not Found')

KeyError: 'Oleg'

In [60]:
# проверка на наличие ключа в словаре
recruit = 'Amanda'

if recruit in salaries:
    print('Значение для ключа уже существует')
else:
    print('Добавляю новый ключ')
    salaries[recruit] = 2200

print(salaries)

Добавляю новый ключ
{'John': 1200, 'Mary': 2000, 'Steven': 1000, 'Liza': 1500, 'Amanda': 2200}


In [61]:
# Можно использовать метод setdefault
# setdefault не изменяет значение, если ключ уже был в словаре

salaries.setdefault('Mary', 3000)
salaries
# salaries.setdefault('Paul', 3000)
# salaries

{'John': 1200, 'Mary': 2000, 'Steven': 1000, 'Liza': 1500, 'Amanda': 2200}

In [62]:
# перейдем к более сложному словарю
staff_dict = {
    'Robert': {'salary': 800, 'bonus': 200}, 
    'Jane': {'salary': 200, 'bonus': 300}, 
    'Liza': {'salary': 1300, 'bonus': 200}, 
    'Richard': {'salary': 500, 'bonus': 1200}
}

In [63]:
staff_dict['Robert']['salary']

800

In [64]:
staff_dict['Oleg'] = {'salary': 1000000, 'bonus': 300}
staff_dict

{'Robert': {'salary': 800, 'bonus': 200},
 'Jane': {'salary': 200, 'bonus': 300},
 'Liza': {'salary': 1300, 'bonus': 200},
 'Richard': {'salary': 500, 'bonus': 1200},
 'Oleg': {'salary': 1000000, 'bonus': 300}}

In [65]:
# получаем только ключи/значения из словаря (очень пригодиться в циклах)
# print(staff_dict.keys())
# print(staff_dict.values())
print(staff_dict.items())

# print(list(staff_dict.keys()))
# print(list(staff_dict.values()))
print(list(staff_dict.items()))

dict_items([('Robert', {'salary': 800, 'bonus': 200}), ('Jane', {'salary': 200, 'bonus': 300}), ('Liza', {'salary': 1300, 'bonus': 200}), ('Richard', {'salary': 500, 'bonus': 1200}), ('Oleg', {'salary': 1000000, 'bonus': 300})])
[('Robert', {'salary': 800, 'bonus': 200}), ('Jane', {'salary': 200, 'bonus': 300}), ('Liza', {'salary': 1300, 'bonus': 200}), ('Richard', {'salary': 500, 'bonus': 1200}), ('Oleg', {'salary': 1000000, 'bonus': 300})]


In [ ]:
# итерация по словарям
# так бы было без цикла
print("Robert's salary:", staff_dict['Robert']['salary'])
print("Jane's salary:", staff_dict['Jane']['salary'])
print("Richard's salary:", staff_dict['Richard']['salary'])

In [66]:
for person in staff_dict:
    print(person)

Robert
Jane
Liza
Richard
Oleg


In [67]:
for key in staff_dict.keys():
    print(key)

Robert
Jane
Liza
Richard
Oleg


In [68]:
for value in staff_dict.values():
    print(value)

{'salary': 800, 'bonus': 200}
{'salary': 200, 'bonus': 300}
{'salary': 1300, 'bonus': 200}
{'salary': 500, 'bonus': 1200}
{'salary': 1000000, 'bonus': 300}


In [69]:
for key, value in staff_dict.items():
    print(key, value)

Robert {'salary': 800, 'bonus': 200}
Jane {'salary': 200, 'bonus': 300}
Liza {'salary': 1300, 'bonus': 200}
Richard {'salary': 500, 'bonus': 1200}
Oleg {'salary': 1000000, 'bonus': 300}


In [70]:
for i, person in enumerate(staff_dict):
    print(i+1, person)

1 Robert
2 Jane
3 Liza
4 Richard
5 Oleg


In [71]:
# используем цикл
for person, info in staff_dict.items():
#     print(person, info)
    print(person, "'s salary: ", info['salary'], sep='')

Robert's salary: 800
Jane's salary: 200
Liza's salary: 1300
Richard's salary: 500
Oleg's salary: 1000000


In [72]:
# добавим уровень з/п
for person, info in staff_dict.items():
#     print(person)
    if info['salary'] > 1000:
        info['status'] = 'above average'
    else:
        info['status'] = 'below average'
#     print(f"{person}'s salary: {info['salary']} ({status})")

staff_dict

{'Robert': {'salary': 800, 'bonus': 200, 'status': 'below average'},
 'Jane': {'salary': 200, 'bonus': 300, 'status': 'below average'},
 'Liza': {'salary': 1300, 'bonus': 200, 'status': 'above average'},
 'Richard': {'salary': 500, 'bonus': 1200, 'status': 'below average'},
 'Oleg': {'salary': 1000000, 'bonus': 300, 'status': 'above average'}}

In [73]:
# функция zip
categories = ['Еда', 'Авто', 'Политика', '346346']
audience = [100, 200, 300]

categories_dict = dict(zip(categories, audience))
print(categories_dict)

{'Еда': 100, 'Авто': 200, 'Политика': 300}


## Dict comprehension
Похоже на  list comprehension

In [74]:
[x**2 for x in range(10)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

In [75]:
{n: n**2 for n in range(10)}

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}

In [76]:
results = [('date', '2018-01-01'), ('counter', '777'), ('visits', 154)]

In [77]:
{metric: value for (metric, value) in results}

{'date': '2018-01-01', 'counter': '777', 'visits': 154}

## Попрактикуемся
Для каждого источника посчитайте ROI (revenue / cost - 1)

In [78]:
results = {
    'vk': {'revenue': 103, 'cost': 98},
    'yandex': {'revenue': 179, 'cost': 153},
    'facebook': {'revenue': 103, 'cost': 110},
    'adwords': {'revenue': 35, 'cost': 34},
    'twitter': {'revenue': 11, 'cost': 24},
}

In [110]:
rio = 0
for _, __ in results.items():
    print(f"{_.capitalize()} ROI is {__['revenue'] / (__['cost'] - 1)}")


Vk ROI is 1.0618556701030928
Yandex ROI is 1.1776315789473684
Facebook ROI is 0.944954128440367
Adwords ROI is 1.0606060606060606
Twitter ROI is 0.4782608695652174


In [107]:
cook_book = {
  'салат': [
     {'ingridient_name': 'сыр', 'quantity': 50, 'measure': 'гр'},
     {'ingridient_name': 'томаты', 'quantity': 20, 'measure': 'гр'},
     {'ingridient_name': 'огурцы', 'quantity': 20, 'measure': 'гр'},
     {'ingridient_name': 'маслины', 'quantity': 10, 'measure': 'гр'},
     {'ingridient_name': 'оливковое масло', 'quantity': 20, 'measure': 'мл'},
     {'ingridient_name': 'салат', 'quantity': 10, 'measure': 'гр'},
     {'ingridient_name': 'перец', 'quantity': 20, 'measure': 'гр'}
    ],
  'пицца': [
     {'ingridient_name': 'сыр', 'quantity': 20, 'measure': 'гр'},
     {'ingridient_name': 'колбаса', 'quantity': 30, 'measure': 'гр'},
     {'ingridient_name': 'бекон', 'quantity': 30, 'measure': 'гр'},
     {'ingridient_name': 'оливки', 'quantity': 10, 'measure': 'гр'},
     {'ingridient_name': 'томаты', 'quantity': 20, 'measure': 'гр'},
     {'ingridient_name': 'тесто', 'quantity': 100, 'measure': 'гр'},   
    ],
  'лимонад': [
     {'ingridient_name': 'лимон', 'quantity': 1, 'measure': 'шт'},
     {'ingridient_name': 'вода', 'quantity': 200, 'measure': 'мл'},
     {'ingridient_name': 'сахар', 'quantity': 10, 'measure': 'гр'},
     {'ingridient_name': 'лайм', 'quantity': 20, 'measure': 'гр'},    
    ]
}

In [166]:
for _, __ in cook_book.items():
    print(f"Для приготовления одной порции {_.capitalize()} необходимо: \n")
    for ___ in __:
        print(f"{___['ingridient_name'].capitalize()} - {___['quantity']} {___['measure']}.")
    print(f"")


Для приготовления одной порции Салат необходимо: 

Сыр - 50 гр.
Томаты - 20 гр.
Огурцы - 20 гр.
Маслины - 10 гр.
Оливковое масло - 20 мл.
Салат - 10 гр.
Перец - 20 гр.

Для приготовления одной порции Пицца необходимо: 

Сыр - 20 гр.
Колбаса - 30 гр.
Бекон - 30 гр.
Оливки - 10 гр.
Томаты - 20 гр.
Тесто - 100 гр.

Для приготовления одной порции Лимонад необходимо: 

Лимон - 1 шт.
Вода - 200 мл.
Сахар - 10 гр.
Лайм - 20 гр.



In [218]:

person = 10
sal_dic= {}
person_dic = {}

for _, __ in cook_book.items():
    for ___ in __:
        if ___['ingridient_name'] not in sal_dic:
            sal_dic[___['ingridient_name']] = [___['quantity'], ___['measure']]
        else:
            sal_dic[___['ingridient_name']][0] += ___['quantity']

if person > 1:
    for _, __ in sal_dic.items():
        person_dic[_] = [__[0] * person, __[1]]
        print(f"{_.capitalize()} - {__[0] * person} {__[1]}")
else:
    for _, __ in sal_dic.items():
        person_dic[_] = [__[0], __[1]]
        print(f"{_.capitalize()} - {__[0] * person} {__[1]}")










Сыр - 700 гр
Томаты - 400 гр
Огурцы - 200 гр
Маслины - 100 гр
Оливковое масло - 200 мл
Салат - 100 гр
Перец - 200 гр
Колбаса - 300 гр
Бекон - 300 гр
Оливки - 100 гр
Тесто - 1000 гр
Лимон - 10 шт
Вода - 2000 мл
Сахар - 100 гр
Лайм - 200 гр
